dom.go

```diff
@@ -69,6 +69,13 @@ type Unmounter interface {
 	Unmount()
 }
 
+// Keyer is an optional interface that a Component can implement in order to
+// uniquely identify the component amongst its siblings. If implemented, all
+// sibling components and HTML must also be keyed.
+type Keyer interface {
+	Key() string
```

Comment:
Add docstring:

```
> Key returns the key value that uniquely identifies this component amongst its siblings.
```

Reply:
Why is a `Keyer` using a `string` key and not an `interface{}` key? I think that cases like this will be very common, and `interface{}` equality works just the same for us internally as `string` equality would (i.e. we don't need to convert to a string).

Reply:
It could be an `interface{}`, string was just the most sensible concrete type to me.  I'm okay with changing it though.

Reply:
Yes, let's change it

Reply:
I'll have to try and document very clearly that keys for all siblings must be of the same type, and that the returned value's type must be comparable, maybe include a link to https://golang.org/ref/spec#Comparison_operators

Reply:
Does the difficulty in documenting this give you any pause, or do you still think that `interface{}` is the correct return here?

Reply:
I am still confident that `interface{}` is a better choice than `string`

I think that saying:

```
// Key returns a value that will be compared against sibling key values for ordering purposes.
```

Is plenty enough. I don't think we need to even link to the spec about comparison operators.

If an `interface{}` value is not comparable to another `interface{}` value, nothing bad will happen (i.e. no panic) aside from the comparison becoming false / not working as the user may incorrectly expect. For example: https://play.golang.org/p/7uK4AQbER-

Anecdotally, although not exactly the same situation, the new `sync.Map` type also doesn't document requirements for keys: https://golang.org/pkg/sync/#Map.Load

Reply:
We don't strictly compare different key values, just look them up, and check for conflicts.  That said, a value that is invalid as a map key will panic: https://play.golang.org/p/HjI8bscDdz.

And `sync.Map` is a rather different scenario - it's obvious that you must adhere to `map` semantics there, whereas here, the fact that the key is stored in a map internally is an implementation detail, and is not obvious to the user.

Reply:
Right, I see, I had forgotten that. How about just:

```
// Key returns a value that will be used for ordering purposes. The type must be valid as
// a map key, or a panic will occur.
```

Clear and to the point. Also, we should make sure to add something like:

```
_ = map[interface{}]struct{}{key: struct{}{}}
```

At any location where we receive a map key from the user of Vecty, as this would ensure they get a stack trace that shows exactly the location where they passed the invalid key. Presumably the compiler will optimize this out completely (but I didn't check, and it doesn't matter).

Reply:
> Also, we should make sure to add something like:

I don't think we bother with this, components can implement `Keyer` however they like anyway, so would be of limited utility.

Reply:
Fine for now, I'll think about it more.